op-batcher/batcher/driver.go

```diff
@@ -373,22 +366,23 @@ func (l *BatchSubmitter) calculateL2BlockRangeToStore(syncStatus eth.SyncStatus)
 		return eth.BlockID{}, eth.BlockID{}, errors.New("empty sync status")
 	}
 
+	lastStoredBlock := l.state.LastStoredBlock()
 	// Check last stored to see if it needs to be set on startup OR set if is lagged behind.
 	// It lagging implies that the op-node processed some batches that were submitted prior to the current instance of the batcher being alive.
-	if l.lastStoredBlock == (eth.BlockID{}) {
-		l.Log.Info("Starting batch-submitter work at safe-head", "safe", syncStatus.SafeL2)
-		l.lastStoredBlock = syncStatus.SafeL2.ID()
-	} else if l.lastStoredBlock.Number < syncStatus.SafeL2.Number {
-		l.Log.Warn("Last submitted block lagged behind L2 safe head: batch submission will continue from the safe head now", "last", l.lastStoredBlock, "safe", syncStatus.SafeL2)
-		l.lastStoredBlock = syncStatus.SafeL2.ID()
+	if lastStoredBlock == (eth.BlockID{}) {
+		l.Log.Info("Resuming batch-submitter work at safe-head", "safe", syncStatus.SafeL2)
+		lastStoredBlock = syncStatus.SafeL2.ID()
+	} else if lastStoredBlock.Number < syncStatus.SafeL2.Number {
+		l.Log.Warn("Last submitted block lagged behind L2 safe head: batch submission will continue from the safe head now", "last", lastStoredBlock, "safe", syncStatus.SafeL2)
+		lastStoredBlock = syncStatus.SafeL2.ID()
```

Comment:
Need to think carefully about whether setting this here is safe / makes sense now. Seems like `pruneSafeBlocks` and `calculateL2BlockRangeToStore` might be stepping on each others toes, maybe they should be combined into one function.

Reply:
This is my only remaining concern about this PR --where `calculateL2BlockRangeToStore` had been modifying the state, it now appears to first consult the state, but then only modifies and returns a local. The two questions I would have are:
* Do callers to this function expect to be setting a value? If so, they're now broken.
* Do other methods have a way of setting the state's LastStoredBlock? If not, the variable isn't functional here.

I would *hope* either of these issues would be discovered via existing tests, but you could also try tracing references back from this function.

Reply:
Pushed another commit to clarify this [d2a77a4](https://github.com/ethereum-optimism/optimism/pull/12981/commits/d2a77a4b90b0d66f6bd60dc6786b55fab310622e). There is only one caller of this function -- it doesn't expect to be setting a value, and I have clarified the godoc to that effect. It's about loading blocks, and now the "last stored block" really is the last stored block (modified via the channel manager), not a separate variable that we have to work to keep in sync. 

It makes sense to me that `calculateL2BlockRangeToStore` is a "view" function.